Context

The iQueue for Operating Rooms is a software product to free up capacity in operating roomsand create a much more transparent and surgeon-centric process for measuring OperatingRoom (OR) utilization.

A typical hospital will have multiple ORs, e.g. 10 rooms and they are staffed to carry out surgical procedures Monday to Friday. OR management has utilized a system called Block Scheduling to arrange and coordinate their surgeons to schedule their surgeries smoothly. The idea is that a room is typically reserved for a certain surgeon, or a group of them on a certain day-of-week.
E.g. Dr. Johnson gets Room 1 on every Monday, the Thoracic group gets Room 2 on everyWednesday etc. It sometimes can be a bit more nuanced, e.g. the frequency could be bi-weekly or others, and sometimes people may give half day blocks instead of full day blocks depending on the typical case length and volume. Also, many hospitals may assign some open time in some rooms on some days, e.g. every Monday and Wednesday Room 5 is open and it follows FCFS. The benefit of the Block Scheduling system is that it provides some certainty for surgeons and other people involved to carry out a surgical case successfully. The problem with it is that sometimes surgeons cannot fill their blocks well with the cases, e.g. they go on vacations or attend conferences or simply have a slower season etc. 

One of the modules within the iQueue for OR is called Exchange, which is a marketplace for ORtime. Basically if Dr. Johnson is going to a conference in 5 weeks and knows that he will not need to utilize that Monday block then he can release the time and other people can have the visibility of that released time and may request it in order to utilize the OR resource. In the
hospital setting, surgeons, or their schedulers will initiate such requests or releases, and then the OR schedulers will review such activities and make sure they are legit and respond and/or update their internal system.



Assignment
You work at iQueue for Operating Rooms as a Data Analyst. Recently a new customer just wentlive with our software. They have adopted the Exchange module and their executive sponsorsfor our product wants to know how the implementation has been.

The customer success team has a few questions:

●How many minutes of requests, transfers and releases have been completed? The marketplace is successful if there's a good amount of activities from both sides.

●What is the denial rate and response time?  The marketplace is easy to use iftransactions are both quick and easy.

In order to get a more complete picture, the customer success team has tasked you to help answer 

​how effective the marketplace is ​by answering these questions along with anyadditional findings you can provide. 

So given the attached data for the Exchange transactionhistory, your job here is to create a narrative for the customer success team and the executivesponsors. 

You need to think through how you want to define effectiveness, e.g. maybe it’s onlyeffective if the time released are requested etc.. 

Also think through the secondary metrics, e.g.
what can the team track and encourage in order to increase the effectiveness etc.?

The exercise may include the following:

●Prepare a short executive summary for your findings and potential actions for the executive sponsors at hospitals.

●Prepare a dashboard that tracks certain metrics that can tell your narrative on how effective the marketplace is.

●Prepare a short note suggesting to the internal team on what additional data needs to be pulled and investigated, if any



The avaiable data set consists of 11 attributes. 


Data Dictionary

transaction_id: A uniquely generated serial number for each individual transactionon the Exchange platform.

parent_transaction_id: The transaction id of the parent transaction or responses (approve,deny, mark updated).

Action
The type of transaction: 
1.RELEASE - A release of existing block time 
2.MARK_UPDATED - The parent transaction has beensuccessfully updated in the hospital’s EHR 
3.REQUEST - A request for open time 
4.APPROVE_REQUEST - The parent request has beenapproved 
5.DENY_REQUEST - The parent request has been denied 
6.TRANSFER - A transfer of time from one person to another 
7.APPROVE_TRANSFER - The parent transfer has beenapproved
8.DENY_TRANSFER - The parent transfer has been denied

scheduler: the person creating the transaction

surgeon:The surgeon for whom the transaction is made.

created_datetime: Date and time when the transaction was created.

snapshot_date: The date of the block associated with the transaction.

start_time: The start time of the block associated with the transaction.

end_time: The end time of the block associated with the transaction.

room_name:The hospital room of the transaction.

location:The hospital location of the transaction.



In [1]:
import pandas as pd #importing required libraries
import numpy as np
import sqlite3 as sql
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
#Create a connection to database file
database = 'LeanTaaSTestDB.db'
connection = sql.connect(database)

In [3]:
query = '''SELECT * FROM exchange_transactions'''

In [4]:
df = pd.read_sql_query(query, connection)
df.head()

transaction_id             parent_transaction_id  \
0  2c91808267d2ea5301682f73e9a1093e                              None   
1  2c91808267d2ea5301683984e52b0bfb                              None   
2  2c91808267d2ea53016839855a480bfc                              None   
3  2c91808267d2eb4b016829281a390786                              None   
4  2c91808267d2eb4b01682977274607cb  2c91808267d2eb4b016829281a390786   

         action       scheduler           surgeon            created_datetime  \
0       RELEASE  Jessie Shelley        Bryan Neer  2019-01-15 13:52:32.000000   
1      TRANSFER   Raymond Smith  Charles Mcdonald  2019-01-17 12:47:17.000000   
2      TRANSFER   Raymond Smith  Charles Mcdonald  2019-01-17 12:47:47.000000   
3       RELEASE   Kimberly Bugg      Joann Hinson  2019-01-14 08:32:01.000000   
4  MARK_UPDATED      Dora Smith      Joann Hinson  2019-01-14 09:58:21.000000   

                snapshot_date start_time  end_time   room_name location  
0  2019-07-01 00:00:00.000000   07:30:00  19:00:00  MRAH OR 03   CENTER  
1  2019-06-20 00:00:00.000000   07:30:00  19:00:00  MRAH OR 06   CENTER  
2  2019-06-20 00:00:00.000000   07:30:00  19:00:00     MNOR 36   CENTER  
3  2019-06-12 00:00:00.000000   08:50:00  19:00:00     MNOR 25   CENTER  
4  2019-06-12 00:00:00.000000   08:50:00  19:00:00     MNOR 25   CENTER

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6965 entries, 0 to 6964
Data columns (total 11 columns):
transaction_id           6965 non-null object
parent_transaction_id    4176 non-null object
action                   6965 non-null object
scheduler                6965 non-null object
surgeon                  6965 non-null object
created_datetime         6965 non-null object
snapshot_date            6965 non-null object
start_time               6965 non-null object
end_time                 6965 non-null object
room_name                4376 non-null object
location                 6965 non-null object
dtypes: object(11)
memory usage: 598.6+ KB


In [6]:
# Total transactions updated
percentage_of_total_parent_transaction_id = df.parent_transaction_id.count() / df.transaction_id.count()*100
print(percentage_of_total_parent_transaction_id)

59.95692749461594


Out of the total transactions 6965 made, 4176 (59.9%) of the transactions has been marked as parent transactions or responses.

In [21]:
#Type of transactions:action
df.action.value_counts()


MARK_UPDATED        2576
REQUEST             1324
RELEASE             1201
APPROVE_REQUEST     1139
TRANSFER             264
APPROVE_TRANSFER     257
DENY_REQUEST         174
DENY_RELEASE          23
DENY_TRANSFER          7
Name: action, dtype: int64

In [11]:
df.action.unique()

array(['RELEASE', 'TRANSFER', 'MARK_UPDATED', 'APPROVE_TRANSFER',
       'REQUEST', 'DENY_REQUEST', 'APPROVE_REQUEST', 'DENY_RELEASE',
       'DENY_TRANSFER'], dtype=object)

In [18]:
# Percentage 
total_transaction =  df.action.value_counts()/ df.action.count()*100
print(total_transaction)

MARK_UPDATED        36.984925
REQUEST             19.009332
RELEASE             17.243360
APPROVE_REQUEST     16.353195
TRANSFER             3.790380
APPROVE_TRANSFER     3.689878
DENY_REQUEST         2.498205
DENY_RELEASE         0.330223
DENY_TRANSFER        0.100503
Name: action, dtype: float64


There has been made 1324 requests for open time, 1201 transacation were released and 264 transactions were trasferred from one person to another.

●What is the denial rate and response time?  The marketplace is easy to use if transactions are both quick and easy.

In [30]:
#Three categories from attribute'action' seem to be the total denied transactions.
#I am aggregating the three denials to calculate the total denial rate.
denial = df[df.action.isin(['DENY_RELEASE', 'DENY_REQUEST','DENY_TRANSFER'  ])].shape[0]
print(denial)

204


In [31]:
#Denial percentage
Denial_rate = denial / df.action.count()*100
print(denial_rate)

2.9289303661162958


The denial rate would be the 2.92% which is sum of total denials (DENY_REQUEST, DENY_RELEASE and DENY_TRANSFER)

In [37]:
#The response time would be the date and time when the transaction was created subtracted from the start time of the block associated with the transaction 
#Converting each of the time field to pd's datetime 
df['start_time']= pd.to_datetime(df.start_time)
df['created_datetime']= pd.to_datetime(df.created_datetime)

In [38]:
df['response_time']= df['start_time']- df['created_datetime']

In [39]:
df.groupby(['response_time']).agg(sum)

transaction  denial
response_time                         
87 days 22:56:08           0.0       0
87 days 22:56:13           0.0       0
88 days 04:50:08           0.0       0
89 days 00:30:17           0.0       0
89 days 01:04:14           0.0       0
89 days 03:09:50           0.0       0
89 days 03:09:54           0.0       0
89 days 04:04:06           0.0       0
89 days 09:11:40           0.0       0
89 days 09:11:54           0.0       0
91 days 19:00:57           0.0       0
91 days 19:01:15           0.0       0
91 days 19:49:31           0.0       0
91 days 20:36:12           0.0       0
91 days 21:11:12           0.0       0
91 days 21:27:16           0.0       0
91 days 21:42:22           0.0       0
91 days 21:47:14           0.0       0
91 days 22:16:18           0.0       0
91 days 22:27:35           0.0       0
91 days 22:37:09           0.0       0
91 days 23:26:17           0.0       0
91 days 23:35:46           0.0       0
92 days 00:30:28           0.0       0
92 days 03:09:08           0.0       0
92 days 03:23:55           0.0       0
92 days 05:53:49           0.0       0
92 days 05:54:01           0.0       0
92 days 06:27:56           0.0       0
92 days 08:25:34           0.0       0
...                        ...     ...
304 days 22:10:59          0.0       0
304 days 23:29:47          0.0       0
304 days 23:34:26          0.0       0
304 days 23:48:37          0.0       0
304 days 23:54:19          0.0       0
304 days 23:58:53          0.0       0
305 days 04:06:31          0.0       0
305 days 04:06:54          0.0       0
305 days 04:19:30          0.0       0
305 days 18:01:31          0.0       0
305 days 18:53:58          0.0       0
305 days 19:41:14          0.0       0
306 days 02:38:06          0.0       0
308 days 18:47:46          0.0       0
308 days 22:15:49          0.0       0
308 days 22:32:59          0.0       0
309 days 01:27:58          0.0       0
309 days 01:28:06          0.0       0
309 days 01:28:27          0.0       0
309 days 01:31:12          0.0       0
309 days 18:42:13          0.0       0
309 days 18:42:43          0.0       0
310 days 20:45:21          0.0       0
310 days 21:01:47          0.0       0
311 days 16:30:22          0.0       0
311 days 17:37:28          0.0       0
311 days 23:43:35          0.0       0
311 days 23:53:10          0.0       0
312 days 22:51:39          0.0       0
313 days 00:17:59          0.0       0

[6825 rows x 2 columns]

In [50]:
#Calculate the minimum, maximum and average response time
print('Average:',np.mean(df['response_time']))
print('Minimum:', np.min(df['response_time']))
print('Maximum:', np.max(df['response_time']))


Average: 164 days 00:52:31.927351
Minimum: 87 days 22:56:08
Maximum: 313 days 00:17:59


In [ ]:
The response time seems to be 163 days 52 minutes and 31 seconds on average.